diff --git a/arch/arm64/boot/dts/renesas/r9a09g011gbg-evaluation-board.dts b/arch/arm64/boot/dts/renesas/r9a09g011gbg-evaluation-board.dts
index 78ba5f5..3fd5e75 100755
--- a/arch/arm64/boot/dts/renesas/r9a09g011gbg-evaluation-board.dts
+++ b/arch/arm64/boot/dts/renesas/r9a09g011gbg-evaluation-board.dts
@@ -22,6 +22,11 @@
         stdout-path = "serial0:115200n8";
     };
 
+    memory@7C000000 {
+        device_type = "memory";
+        reg = <0x0 0x7C000000 0x0 0x4000000>;
+    };
+    
     memory@180000000 {
         device_type = "memory";
         reg = <0x1 0x80000000 0x0 0x80000000>;
@@ -32,6 +37,11 @@
         #size-cells = <2>;
         ranges;
 
+        rtos_rpmsg_reserved: rtos_rpmsg@7C000000 {
+             no-map;
+             reg = <0x0 0x7C000000 0x0 0x4000000>;
+        };
+        
         /* global autoconfigured region for contiguous allocations */
         global_cma: linux,cma@1B8000000 {
             compatible = "shared-dma-pool";
@@ -64,6 +74,45 @@
         memory-region = <&image_buf0>;
     };
 
+    rsctbl: rsctbl@7FDD0000 {
+        compatible = "shm_uio";
+        reg = <0x0 0x7FDD0000 0x0 0x20000>;
+        no-map;
+    };
+    
+    shm1:shm@7FE00000 {
+        compatible = "shm_uio";
+        reg = <0 0x7FE00000 0 0x00040000>;
+    };
+
+    shm0:shm@7FE40000 {
+        compatible = "shm_uio";
+        reg = <0 0x7FE40000 0 0x00040000>;
+    };
+
+    /* VRING area for OpenAMP */
+    vring1:vring@7FE80000 {
+        compatible = "vring_uio";
+        reg = <0 0x7FE80000 0 0x00040000>;
+    };
+
+    vring0:vring@7FEC0000 {
+        compatible = "vring_uio";
+        reg = <0 0x7FEC0000 0 0x00040000>;
+    };
+
+    gic0:gic0@82000000 {
+        compatible = "shm_uio";
+        reg = <0 0x82000000 0 0x00080000>;
+        ipi = <8>;
+    };
+    
+    gic1:gic1@82000000 {
+        compatible = "shm_uio";
+        reg = <0 0x82000000 0 0x00080000>;
+        ipi = <9>;
+    };
+
     vcc_sdhi0: regulator-vcc-sdhi0 {
         compatible = "regulator-fixed";
 
