diff --git a/lib/include/openamp/virtqueue.h b/lib/include/openamp/virtqueue.h
index ad24749..ea63948 100755
--- a/lib/include/openamp/virtqueue.h
+++ b/lib/include/openamp/virtqueue.h
@@ -57,6 +57,11 @@ struct virtqueue_buf {
 	int len;
 };
 
+struct vq_desc_extra {
+	void *cookie;
+	uint16_t ndescs;
+};
+
 struct virtqueue {
 	struct virtio_device *vq_dev;
 	const char *vq_name;
@@ -99,10 +104,7 @@ struct virtqueue {
 	 * Other fields in this structure are not used currently.
 	 */
 
-	struct vq_desc_extra {
-		void *cookie;
-		uint16_t ndescs;
-	} vq_descx[0];
+	struct vq_desc_extra vq_descx[0];
 };
 
 /* struct to hold vring specific information */
